### PR TITLE
Disk space check / download precentage correction / shutdown process

### PR DESCRIPTION
**Pull request by Tostifrosti**
This pull request contains a few fixes & improvements:
- Added interpreter pattern in startup launcher (this makes it easier to add stuff in the future)
- Disk space check before downloading instructions
- Download percentage corrections
- User cancellation improved
- Installer will keep the downloads after user has interupted the download
- Overall improvements (added 'using' to dispose some objects after they're done)
- Removed some unnecessary usings (always good to keep it nice and tidy :D)
- Fixed the shutdown process by stopping all threads before shutdown. (instead of killing the entire application with: `Application.Current.Shutdown()`)

Hopefully you guys will find these changes good enough to be implemented ;)
Feel free to comment on any changes

### CODE REVIEW

**Comment by JAJames** [comment]
When testing this, I observed the following when installing a patch:
image

It also seems that anytime I launch the launcher, it prompts to update to the latest version (even if it already is):
image

And unrelated to your changes, the above prompt shows the version number being cut off because the display field is too small -- we need to expand this.

**Comment by JAJames** [comment]
There's also some other issues, but aren't necessarily related to this PR and shouldn't be considered "blocking" (similar to the version text field). I'll edit this comment as more issues are discovered, so that they can be converted into issues later.

If a mirror is unable to deliver files (i.e: setup a proxy, and start refusing connections mid-install for a specific mirror), the launcher will continue attempting to use that mirror indefinitely. We need to change this behavior such that any mirror which fails the retry strategy is removed from the valid mirror list, so that it's not used for future downloads by the client.

File: RXPatchLib/DirectoryPatcher.cs [hunk 1]
```diff
@@ -309,30 +309,65 @@ public DirectoryPatcher(XdeltaPatcher filePatcher, string targetPath, string bac
...
             }
 
             // Deserialize JSON data from headerFileContents
             List<FilePatchInstruction> instructions = JsonConvert.DeserializeObject<List<FilePatchInstruction>>(headerFileContents);
 
             // Initialize progress-related variables
-            var progress = new DirectoryPatchPhaseProgress();
             var paths = instructions.Select(i => Path.Combine(_targetPath, i.Path));
             var sizes = paths.Select(p => !File.Exists(p) ? 0 : new FileInfo(p).Length);
-            progress.SetTotals(instructions.Count, sizes.Sum());
```
Comment (JAJames) [inline RXPatchLib/DirectoryPatcher.cs]: I suspect that it's the changes in this area that's causing the inaccurate progress percentages. I'll try to expand on this thought later; unfortunately I scrapped my previous review of this area.

**Comment by Tostifrosti** [comment]
Sorry for the inactivity. I’ve been quite busy and didn’t have much time to work on those bugs. I’ll try my best to fix all of this in the coming 2 months.

**Pull request closed without merging**